model/PciRegistry.php

```diff
@@ -50,267 +44,378 @@ class PciRegistry extends ConfigurableService
-        $deleted = false;
-        $fs = $this->getFileSystem();
+
+    /**
+     * Unregister files by removing them from FileSystem
+     *
+     * @param PciModel $pciModel
+     * @param $files
+     * @return bool
+     * @throws \common_Exception
+     */
+    protected function unregisterFiles(PciModel $pciModel, $files)
+    {
+        $deleted = true;
+        $filesystem = $this->getFileSystem();
         foreach ($files as $relPath) {
-            $fileId = $this->getPrefix($id, $version).$relPath;
-            if($fs->has($fileId)){
-                $deleted &= $fs->delete($fileId);
-            }else{
-                throw new \common_Exception('File does not exists in the filesystem : ' . $relPath);
+            $fileId = $this->getPrefix($pciModel) . $relPath;
+            if (!$filesystem->has($fileId)) {
+                throw new \common_Exception('File does not exists in the filesystem: ' . $relPath);
             }
+            $deleted = $filesystem->delete($fileId);
         }
         return $deleted;
     }
-    
-    protected function getMap(){
-        return \common_ext_ExtensionsManager::singleton()->getExtensionById('qtiItemPci')->getConfig(self::CONFIG_ID);
-    }
-    
-    protected function setMap($map){
-        \common_ext_ExtensionsManager::singleton()->getExtensionById('qtiItemPci')->setConfig(self::CONFIG_ID, $map);
+
+    /**
+     * Return the last version of a PCI track by $typeIdentifier
+     *
+     * @param $typeIdentifier
+     * @return mixed|null
+     */
+    protected function getLatestVersion($typeIdentifier)
+    {
+        return $this->get($typeIdentifier);
     }
-    
-    public function getLatestVersion($typeIdentifier){
+
+    /**
+     * Check if PCI exists into self::CONFIG_ID mapping
+     *
+     * @param PciModel $pciModel
+     * @return bool
+     */
+    public function exists(PciModel $pciModel)
+    {
         $pcis = $this->getMap();
-        if(isset($pcis[$typeIdentifier])){
-            $pci = $pcis[$typeIdentifier];
-            end($pci);
-            return key($pci);
+
+        if (empty($pcis) || !isset($pcis[$pciModel->getTypeIdentifier()])) {
+            return false;
         }
-        return null;
+
+        $version = $pciModel->getVersion();
+        if ($pciModel->getVersion()===null) {
+            $version = $this->getLatestVersion($pciModel->getTypeIdentifier());
+        }
+
+        if ($version!==null) {
+            return (isset($pcis[$pciModel->getTypeIdentifier()][$version]));
+        }
+
+        return false;
     }
-    
-    public function exists($typeIdentifier, $version = ''){
+
+    /**
+     * Get a PCI from identifier/version
+     * $versionStrict to search exactly the given $version
+     *
+     * @param $identifier
+     * @param null $version
+     * @param bool $versionStrict
+     * @return $this|null
+     */
+    public function get($identifier, $version=null, $versionStrict=false)
```

Comment:
is the thrid arg versionStrict really necessary ? I am wondering about the different use cases....
If the version is given, then it should get that specific version otherwise nothing.
If no given version is given, we return the latest one.